# Homework #5 Bodie Franklin

In [1]:
%%time
import bs4 as bs
import urllib.request

Wall time: 161 ms


## Question 1  

In [40]:
%%time
reviews = []
review_link = []

the_dark_knight = "https://www.imdb.com/title/tt0468569/reviews?ref_=tt_urv"
sauce= urllib.request.urlopen(the_dark_knight).read()
soup = bs.BeautifulSoup(sauce,'lxml')
    
for a in soup.find_all('a',href=True):
        reviews.append(a['href'])

Infinite = "https://www.imdb.com/title/tt6654210/reviews?ref_=tt_urv"
sauce= urllib.request.urlopen(Infinite).read()
soup = bs.BeautifulSoup(sauce,'lxml')
    
for a in soup.find_all('a',href=True):
        reviews.append(a['href'])
        
Awake = "https://www.imdb.com/title/tt10418662/reviews?ref_=tt_urv"
sauce= urllib.request.urlopen(Awake).read()
soup = bs.BeautifulSoup(sauce,'lxml')
    
for a in soup.find_all('a',href=True):
        reviews.append(a['href'])
        

F9 = "https://www.imdb.com/title/tt5433138/reviews?ref_=tt_urv"
sauce= urllib.request.urlopen(F9).read()
soup = bs.BeautifulSoup(sauce,'lxml')
    
for a in soup.find_all('a',href=True):
        reviews.append(a['href'])
                

hitman_wife = "https://www.imdb.com/title/tt8385148/reviews?ref_=tt_urv"
sauce= urllib.request.urlopen(hitman_wife).read()
soup = bs.BeautifulSoup(sauce,'lxml')
    
for a in soup.find_all('a',href=True):
        reviews.append(a['href'])
            

    
PartialWord ="/review/"
clean_reviews = list(set([word for word in reviews if word.startswith(PartialWord)]))


for i in (clean_reviews):
    review_link.append('https://www.imdb.com/'+i+'?ref_=tt_urv')

print("For movie genre Action,We have",len(review_list),"reviews for our dataset")

For movie genre Action,We have 125 reviews for our dataset
Wall time: 2.9 s


In [41]:
for x in review_link:
    print(x)

https://www.imdb.com//review/rw7018890/?ref_=tt_urv
https://www.imdb.com//review/rw7013509/?ref_=tt_urv
https://www.imdb.com//review/rw7069905/?ref_=tt_urv
https://www.imdb.com//review/rw1911969/?ref_=tt_urv
https://www.imdb.com//review/rw7049211/?ref_=tt_urv
https://www.imdb.com//review/rw5478826/?ref_=tt_urv
https://www.imdb.com//review/rw7041358/?ref_=tt_urv
https://www.imdb.com//review/rw7048732/?ref_=tt_urv
https://www.imdb.com//review/rw7060606/?ref_=tt_urv
https://www.imdb.com//review/rw7056223/?ref_=tt_urv
https://www.imdb.com//review/rw7028159/?ref_=tt_urv
https://www.imdb.com//review/rw7044385/?ref_=tt_urv
https://www.imdb.com//review/rw6475511/?ref_=tt_urv
https://www.imdb.com//review/rw7016949/?ref_=tt_urv
https://www.imdb.com//review/rw7061799/?ref_=tt_urv
https://www.imdb.com//review/rw7045873/?ref_=tt_urv
https://www.imdb.com//review/rw7016350/?ref_=tt_urv
https://www.imdb.com//review/rw7066823/?ref_=tt_urv
https://www.imdb.com//review/rw7037837/?ref_=tt_urv
https://www.

Print out above shows the 125 permalinks that will be used for this exercise.  Please note that these movies are from the action genre. I used The Dark Knight, Infinite, Awake ,F9 and the hitman's wife.

# Question 2

## Part A

In [42]:
%%time
reviews_text = []


for i in review_link:
    link = i
    sauce= urllib.request.urlopen(link).read()
    soup = bs.BeautifulSoup(sauce,'lxml')
    
    for div in soup.find_all('div',class_="text show-more__control"):
        reviews_text.append(div.text)

Wall time: 1min 5s


In [45]:
len(reviews_text)


125

Checking if every link has a review and it does. 

## Part B

In [65]:
%%time
tokenized_words = []

import itertools
import nltk
from nltk.tokenize import word_tokenize
import pprint
import re

def preprocess(doc):
    sentences = nltk.sent_tokenize(doc)
    sentences = [nltk.word_tokenize(sent) for sent in sentences]
    sentences = [nltk.pos_tag(sent) for sent in sentences]
    return sentences


for i in reviews_text:
    tokenized_words.append(preprocess(i))
    


Wall time: 1.05 s


In [66]:
%%time
result = []
grammar = "NP: {<DT>?<JJ>*<NN>}"
NPChunker = nltk.RegexpParser(grammar)

for i in tokenized_words:
    result.append(NPChunker.parse(i[0]))


Wall time: 10.1 ms


In [67]:
result

[Tree('S', [('I', 'PRP'), ('saw', 'VBD'), Tree('NP', [('the', 'DT'), ('trailer', 'NN')]), ('for', 'IN'), Tree('NP', [('this', 'DT'), ('movie', 'NN')]), ('not', 'RB'), ('too', 'RB'), ('long', 'RB'), ('ago', 'RB'), ('and', 'CC'), ('was', 'VBD'), ('expecting', 'VBG'), Tree('NP', [('something', 'NN')]), ('great', 'JJ'), ('.', '.')]),
 Tree('S', [('Always', 'NNS'), ('it', 'PRP'), ('seems', 'VBZ'), ('in', 'IN'), ('these', 'DT'), ('types', 'NNS'), ('of', 'IN'), ('films', 'NNS'), (',', ','), ('there', 'EX'), ("'s", 'VBZ'), Tree('NP', [('the', 'DT'), ('irritating', 'NN')]), Tree('NP', [('angsty', 'JJ'), ('teenager', 'NN')]), (',', ','), Tree('NP', [('the', 'DT'), ('smart', 'JJ'), ('little', 'JJ'), ('kid', 'NN')]), ('who', 'WP'), ('holds', 'VBZ'), ('the', 'DT'), ('secret', 'JJ'), ('and', 'CC'), ('the', 'DT'), ('incredibly', 'RB'), ('annoying', 'VBG'), ('Mom', 'NNP'), ('who', 'WP'), ('acts', 'VBZ'), ('as', 'IN'), ('though', 'IN'), ('she', 'PRP'), ("'s", 'VBZ'), ('going', 'VBG'), ('to', 'TO'), ('s

## Part C

 Based on the output above, it seems that step B was surprising successful and there was no need to add to the lexicon etc. In my opinion, the dark knight and joker were classified correctly. This makes sense since they are names of characters.

# Question 3

In [69]:
result

[Tree('S', [('I', 'PRP'), ('saw', 'VBD'), Tree('NP', [('the', 'DT'), ('trailer', 'NN')]), ('for', 'IN'), Tree('NP', [('this', 'DT'), ('movie', 'NN')]), ('not', 'RB'), ('too', 'RB'), ('long', 'RB'), ('ago', 'RB'), ('and', 'CC'), ('was', 'VBD'), ('expecting', 'VBG'), Tree('NP', [('something', 'NN')]), ('great', 'JJ'), ('.', '.')]),
 Tree('S', [('Always', 'NNS'), ('it', 'PRP'), ('seems', 'VBZ'), ('in', 'IN'), ('these', 'DT'), ('types', 'NNS'), ('of', 'IN'), ('films', 'NNS'), (',', ','), ('there', 'EX'), ("'s", 'VBZ'), Tree('NP', [('the', 'DT'), ('irritating', 'NN')]), Tree('NP', [('angsty', 'JJ'), ('teenager', 'NN')]), (',', ','), Tree('NP', [('the', 'DT'), ('smart', 'JJ'), ('little', 'JJ'), ('kid', 'NN')]), ('who', 'WP'), ('holds', 'VBZ'), ('the', 'DT'), ('secret', 'JJ'), ('and', 'CC'), ('the', 'DT'), ('incredibly', 'RB'), ('annoying', 'VBG'), ('Mom', 'NNP'), ('who', 'WP'), ('acts', 'VBZ'), ('as', 'IN'), ('though', 'IN'), ('she', 'PRP'), ("'s", 'VBZ'), ('going', 'VBG'), ('to', 'TO'), ('s

I used beaituful soup to pull the permalinks from IMDB for the action genre. I pulled reviews for the movies  The Dark Knight, Infinite, Awake ,F9 and the hitman's wife. I collected 125 reviews through a for loop of the review links and scrapping the permlink pages.